benchmark/src/grpc_helpers.rs

```diff
@@ -162,22 +170,97 @@ fn handle_account_state_response(
     client: &AdmissionControlClient,
-    account_address: &[AccountAddress],
-) -> Result<Vec<(u64, AccountStatus)>> {
-    let future_resps: Result<Vec<_>> = account_address
+    addresses: &[AccountAddress],
+) -> HashMap<AccountAddress, (u64, AccountStatus)> {
+    let futures: Vec<_> = addresses
         .iter()
-        .map(|address| get_account_state_async(client, address))
+        .filter_map(|address| match get_account_state_async(client, *address) {
+            Ok(future) => Some(future),
+            Err(e) => {
+                error!("Failed to send account request: {:?}", e);
+                None
+            }
+        })
         .collect();
-    future_resps?
-        .into_iter()
-        .map(handle_account_state_response)
-        .collect()
+    let future_stream = stream::futures_unordered(futures);
+    // Collect successfully requested account states.
+    let mut states = HashMap::new();
+    for pair_result in future_stream.wait() {
+        match pair_result {
+            Ok((address, future_resp)) => match handle_account_state_response(future_resp) {
+                Ok((sequence_number, status)) => {
+                    debug!(
+                        "Update {:?}'s sequence number to {:?}",
+                        address, sequence_number
+                    );
+                    states.insert(address, (sequence_number, status));
+                }
+                Err(e) => {
+                    error!("Invalid account response for {:?}: {:?}", address, e);
+                }
+            },
+            Err(e) => {
+                error!("Failed to receive account response: {:?}", e);
+            }
+        }
+    }
+    states
+}
+
+/// For each sender account, synchronize its persisted sequence number from validator.
+/// When this sync sequence number equals the account's local sequence number,
+/// all its transactions are committed. Timeout if such condition is never met for all senders.
+/// Return sender accounts' most recent persisted sequence numbers.
+pub fn sync_account_sequence_number(
+    client: &AdmissionControlClient,
+    senders: &[AccountData],
+) -> HashMap<AccountAddress, u64> {
+    let mut num_iters = 0;
+    let mut finish = false;
+    // Invariants for the keys in targets (T), unfinished (U) and finished (F):
+    // (1) T = U union F, and (2) U and F are disjoint.
+    let targets: HashMap<AccountAddress, u64> = senders
+        .iter()
+        .map(|sender| (sender.address, sender.sequence_number))
+        .collect();
+    let mut unfinished: HashMap<AccountAddress, u64> =
+        senders.iter().map(|sender| (sender.address, 0)).collect();
+    let mut finished = HashMap::new();
+
+    while !finish && num_iters < MAX_WAIT_COMMIT_ITERATIONS {
+        let unfinished_addresses: Vec<_> = unfinished.keys().copied().collect();
+        let states = get_account_states(client, &unfinished_addresses);
+        for (address, (sequence_number, _status)) in states.iter() {
```

Comment:
Reply to Sonia:

Switched to FutureUnordered after adopting your suggestion to return (address, response_item) in `get_account_state_async`